src/Chainweb/MinerReward.hs

```diff
@@ -0,0 +1,250 @@
+
+-- -------------------------------------------------------------------------- --
+-- Miner Reward
+
+--  | Miner Reward in Stu
+--
+newtype MinerReward = MinerReward { _minerReward :: Stu }
+    deriving (Show, Eq, Ord, Generic)
+
+minerRewardKda :: MinerReward -> Kda
+minerRewardKda (MinerReward d) = stuToKda d
+
+-- | Calculate miner reward for a block at the given height.
+--
+-- NOTE:
+-- This used to compute the value as @roundTo 8 $ (_kda $ stuToKda m) / n@.
+-- The new caclulcation based on Stu is equivalent for 10 and 20 chains,
+-- except for the pre-last entry in the miner rewards table, namely
+-- @(125538056,0.023999333). However, since this value hasen't yet been used
+-- in any network, we can still change the algorithm.
+--
+blockMinerReward
+    :: ChainwebVersion
+    -> BlockHeight
+    -> MinerReward
+blockMinerReward v h = case M.lookupGE h minerRewards of
+    Nothing -> MinerReward $ Stu 0
+    Just (_, s) -> MinerReward $ divideStu s n
+  where
+    !n = int . order $ chainGraphAt v h
+
+-- -------------------------------------------------------------------------- --
+-- Internal
+-- -------------------------------------------------------------------------- --
+
+-- -------------------------------------------------------------------------- --
+-- Miner Rewards Table
+
+type MinerRewardsTable = M.Map BlockHeight Stu
+
+-- | Rewards table mapping 3-month periods to their rewards according to the
+-- calculated exponential decay over about a 120 year period (125538057 block
+-- heights).
+--
+-- It provides the total reward per block height accross all chains. Use the
+-- 'blockMinerReward' function to obtain the reward for a single block at a
+-- given block height.
+--
+-- Morally this is a property of the Chainweb version, however there is no need
+-- to use value different from what is used on Mainnet on any network.
+--
+-- Mining rewards are between 0 and 24 KDA. Values decrease monotonically over
+-- 125538057 block heights (about 120 years).
+--
+minerRewards :: MinerRewardsTable
+minerRewards = mkMinerRewards
+{-# NOINLINE minerRewards #-}
+
+-- | Compute the miner rewards table.
+--
+-- The indirection from 'minerReward' to 'mkMinerReward' is required because the
+-- HasCallStack constraints prevents this value from being a CAF that gets
+-- cached.
+--
+mkMinerRewards :: HasCallStack => MinerRewardsTable
+mkMinerRewards =
+    case CSV.decode CSV.NoHeader (BL.fromStrict rawMinerRewards) of
+        Left e -> error
+            $ "cannot construct miner rewards table: " <> sshow e
+        Right vs ->
+            let rewards = M.fromList . V.toList . V.map formatRow $ vs
+            in if (minerRewardsHash rewards == expectedMinerRewardsHash)
+                then rewards
+                else error $ "hash of miner rewards table does not match expected hash"
+  where
+    formatRow :: (Word64, CsvDecimal) -> (BlockHeight, Stu)
+    formatRow (!a,!b) = (BlockHeight $ int a, kdaToStu (Kda $ _csvDecimal b))
```

Comment:
I don't think this bang changes anything, as the rows are already parsed. You could add bangs to the RHS, but that's probably unnecessary, because this all gets hashed immediately anyway.